secure_storage: remove experimental status

tomi-font opened the pull request:
The subsystem has been out for a year.
It has been taken into use by various parties and has gone through additional downstream testing.
It is considered stable and mature enough to not be experimental anymore.

tomi-font committed:
secure_storage: remove experimental status

The subsystem has been out for a year.
It has been taken into use by various parties and has gone through
additional downstream testing.
It is considered stable and mature enough to not be experimental
anymore.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

kartben reviewed (changes requested):
Not sure if jumping straight from experimental to stable and skipping unstable is what you want, but I don't necessarily oppose to it if you feel strongly about it and no one else has issues with it.

What I would ask though is that you take some time to fully document API https://docs.zephyrproject.org/api-coverage/latest/home/runner/work/zephyr/zephyr/zephyr/subsys/secure_storage/include/psa/index.html and also make sure that it's properly versioned as well. See https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#stable

dkalowsk reviewed (dismissed):
Very happy to see this!

tomi-font committed:
secure_storage: remove NULL pointer checks

It is the caller's responsibility to ensure pointers are valid.
Save on code size.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
secure_storage: its: improve return codes

Instead of returning storage-related error codes, return ones which
make it clear that it's not about the storage itself.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
secure_storage: its: small language improvements to function comments

Minor tweaks for clarity.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
secure_storage: document psa/error.h

Add Doxygen documentation to everything defined in this header file.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
secure_storage: add version to psa_secure_storage Doxygen group

Align the version with the current latest published specification.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font commented:
> needs to be formalized

How exactly?

> if that.was really your intent and unstable is not desirable for whatever reason

What I'm a bit confused about is that here the API exposed by the subsystem to the application is the [PSA Secure Storage API](https://arm-software.github.io/psa-api/storage/) which is stable and not even defined by Zephyr itself. So in that sense the API was stable from the beginning.
To me the `EXPERIMENTAL` Kconfig option rather meant that the subsystem (implementation) itself was experimental. However the documentation you are linking to only talks about APIs?

We do have function prototypes that can be exposed to and used by end users to provide their own definitions of certain functions, under [subsys/secure_storage/include/internal/zephyr/secure_storage](https://docs.zephyrproject.org/latest/doxygen/html/dir_b251feb5349caf21c27bf417dfd4e083.html).
Do those qualify as an API that could be marked as unstable?
